src/he/pmtud.c

```diff
@@ -0,0 +1,287 @@
+#include "he_internal.h"
+
+#include "pmtud.h"
+
+#ifndef MAX
+#define MAX(x, y) (((x) > (y)) ? (x) : (y))
+#endif
+
+#ifndef MIN
+#define MIN(x, y) (((x) < (y)) ? (x) : (y))
+#endif
+
+static void he_internal_pmtud_change_state(he_conn_t *conn, he_pmtud_state_t state) {
+  if(conn == NULL) {
+    return;
+  }
+  if(conn->pmtud_state_change_cb) {
+    conn->pmtud_state_change_cb(conn, state, conn->data);
+  }
+  conn->pmtud_state = state;
+}
+
+he_return_code_t he_internal_pmtud_send_probe(he_conn_t *conn, uint16_t probe_mtu) {
+  if(!conn) {
+    return HE_ERR_NULL_POINTER;
+  }
+
+  if(conn->state != HE_STATE_ONLINE || conn->pmtud_state == HE_PMTUD_STATE_DISABLED) {
+    return HE_ERR_INVALID_CONN_STATE;
+  }
+
+  if(probe_mtu < MIN_PLPMTU || probe_mtu > MAX_PLPMTU) {
+    return HE_ERR_INVALID_MTU_SIZE;
+  }
+
+  // Craft a probe message
+  uint8_t buf[HE_MAX_WIRE_MTU] = {0};
+  he_msg_ping_t *probe = (he_msg_ping_t *)buf;
+  probe->msg_header.msgid = HE_MSGID_PING;
+
+  // Set identifier
+  uint16_t id = conn->ping_next_id++;
+  probe->id = htons(id);
+
+  // Set payload for the probe message
+  uint16_t payload_size = probe_mtu + sizeof(he_msg_data_t) - sizeof(he_msg_ping_t);
+  probe->length = payload_size;
+
+  conn->pmtud_probe_count++;
+  conn->pmtud_probing_size = probe_mtu;
+
+  // Send it
+  he_return_code_t res = he_internal_send_message(conn, buf, sizeof(he_msg_ping_t) + payload_size);
+  int timeout_ms = PMTUD_PROBE_TIMEOUT_MS;
+  if(res != HE_SUCCESS) {
+    // If he_internal_send_message failed to send the probe message, trigger the pmtud timeout with
+    // a tiny delay
+    timeout_ms = 10;
+  }
+
+  conn->pmtud_probe_pending_id = id;
+
+  // Start the probe timer
+  if(conn->pmtud_time_cb) {
+    conn->pmtud_time_cb(conn, timeout_ms, conn->data);
+  }
+
+  return HE_SUCCESS;
+}
+
+he_return_code_t he_internal_pmtud_handle_probe_ack(he_conn_t *conn, uint16_t probe_id) {
+  if(!conn) {
+    return HE_ERR_NULL_POINTER;
+  }
+  if(conn->pmtud_probe_pending_id != probe_id) {
+    // Receives an outdated probe id, ignore.
+    return HE_SUCCESS;
+  }
+
+  // Reset the pending probe id
+  conn->pmtud_probe_pending_id = 0;
+
+  // Reset the probe count for each ack
+  conn->pmtud_probe_count = 0;
+
+  // Handle the ack based on current state
+  switch(conn->pmtud_state) {
+    case HE_PMTUD_STATE_BASE:
+      // Base MTU confirmed
+      return he_internal_pmtud_base_confirmed(conn);
+    case HE_PMTUD_STATE_SEARCHING:
+      // Current probe acked
+      if(conn->pmtud_probing_size >= MAX_PLPMTU) {
+        // Search completed if the MAX_PLPMTU acked
+        return he_internal_pmtud_search_completed(conn);
+      } else {
+        // Increment probing size and send next probe
+        uint16_t probe_size = conn->pmtud_probing_size;
+        probe_size +=
+            (conn->pmtud_is_using_big_step) ? PMTUD_PROBE_BIG_STEP : PMTUD_PROBE_SMALL_STEP;
+        conn->pmtud_probing_size = MIN(probe_size, MAX_PLPMTU);
```

Comment:
We do not need to update the `conn->pmtud_probing_size` here. We are doing that already in `he_internal_pmtud_send_probe` which seems right

Reply:
Fixed.